### PR TITLE
katib: Add per component packages

### PR DESCRIPTION
**Pull request by discordianfish**
**Description of your changes:**
katib: Add per component packages

To support both katib deployment with bundled mysql and external
databases, split the katib-controller package into individual components.

As suggested by @jlewi, this preserves backwards compatibility by
creating new packages that reference the existing manifests

**Checklist:**
- [x] Unit tests have been rebuilt: 
    1. `cd manifests/tests`
    2. `make generate-changed-only`
    3. `make test`

### CODE REVIEW

**Comment by discordianfish** [comment]
Can't reproduce the error locally. It would be great if we could get https://github.com/kubeflow/kfctl/pull/264 merged so kfctl returns *why* it "couldn't generate kustomization file for component katib-controller-standalone"

@jlewi: That being said, is this in general how you imaging it?

**Comment by jlewi** [comment]
Thanks @discordianfish. A couple suggestions

* Preserve backwards compatibility so we can avoid having to update all the KFDefs in this PR.
  * You can create new packages that pull in the YAML resources from their current location
  * I think as you've pointed out the [kfctl kustomize create logic](https://github.com/kubeflow/kfctl/blob/7341e44068be5a5d4ada4f5fa1b30880f82b6fe5/pkg/kfapp/kustomize/kustomize.go#L1151) is a bit brittle and hard to understand
  * So the pattern I've been following is to leave the existing structure intact (i.e. keep base as I think the logic depends on it)
  * Define new kustomize packages that resource the same "<resource>.yaml" files to avoid duplication
  * Once all the KFDefs can switch over to not relying on kustomize files being created by kfctl
    we can cleanup all the individual applications.

* I've been thinking about how to structure our kustomize packages. I might suggest a layout like the following

  ```
  katib/components
                             /katib-db-manager
                            /katib-controller
                           /katib-db-mysql
  katib/installs
                     /katib-standalone
                    /katib-external-db
  ```

  * The idea is to distinguish between install packages and components/modules with the goal of
    making it easier for end-users to find the packages they want.

**Comment by andreyvelich** [comment]
> * I've been thinking about how to structure our kustomize packages. I might suggest a layout like the following
>   ```
>   katib/components
>                              /katib-db-manager
>                             /katib-controller
>                            /katib-db-mysql
>   katib/installs
>                      /katib-standalone
>                     /katib-external-db
>   ```
>   
>   
>   
>   * The idea is to distinguish between install packages and components/modules with the goal of
>     making it easier for end-users to find the packages they want.

This design lgtm /cc @johnugeorge @gaocegege.

@jlewi What do you think about katib-crds: https://github.com/kubeflow/manifests/tree/master/katib/katib-crds ?
Should it be in independent folder? Do we need additional overlay with Application for the Katib CRD? Or we can we add it `/katib/components/ ` folder too?

**Comment by andreyvelich** [comment]
@discordianfish Thanks for doing this!
I will take a look.

File: katib/installs/katib-external-db/base/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+namespace: kubeflow
+resources:
+  - ../../../components/katib-controller
+  - ../../../components/katib-db-manager
+patchesStrategicMerge:
+  - katib-db-manager-deployment.yaml
+secretGenerator:
```
[the remark this answers is not in the document]
Comment (discordianfish) [reply katib/installs/katib-external-db/base/kustomization.yaml:9]: It contains secrets, so I used the secretGenerator. We could use a configmap for everything except the password but I usually prefer keeping everything in a secret in such case.

File: katib/components/katib-controller/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    newTag: v0.8.0
+    newName: gcr.io/kubeflow-images-public/katib/v1alpha3/katib-ui
+commonLabels:
+  app.kubernetes.io/component: katib
+  app.kubernetes.io/instance: katib-controller
+  app.kubernetes.io/managed-by: kfctl
+  app.kubernetes.io/name: katib-controller
+  app.kubernetes.io/part-of: kubeflow
+configurations:
+  - ../../katib-controller/v3/params.yaml
```
[the remark this answers is not in the document]
Comment (discordianfish) [reply katib/components/katib-controller/kustomization.yaml]: Yeah wasn't sure about how to deal with the v3 stuff or not. I've changed it.

**Comment by discordianfish** [comment]
@andreyvelich ok, addressed your comments. Any thoughts about tests?

**Comment by jlewi** [comment]
Just my oppinion so take it or leave it but I might suggest that instead of doing base and overlays in installs structure it as

```
installs:
   katib-external-db
   katib-standalone/ <no base subdirectory
   katib-ibm
```

Nesting things in "/base" and "/overlays" just seems like it makes things harder to find.

Just my 2 cents though.

For the tests add to [search-dirs](https://github.com/kubeflow/manifests/blob/f0de832bb48b9a82660310cf9b3c706a7359050f/hack/generate_tests.py#L16)

The directory whose kustomizations you want to test. So in this case that's probably the "installs" directory or just specific subdirectories.

The test will generate and check in the expected YAML. This way during PR review inspecting the diff of the expected output will tell the reviewer whether anything changed.

**Comment by discordianfish** [comment]
> Just my oppinion so take it or leave it but I might suggest that instead of doing base and overlays in installs structure it as

Makes sense, I actually did it that way first but wasn't sure about the conventions. I've changed it but I called it `katib-standalone-ibm` instead of just `katib-ibm` which seems more appropriate, agreed?

> For the tests add to search-dirs
Did that and generated the expected YAML.

If everybody is fine with the structure, I can validate the changes on our cluster.

**Comment by jlewi** [comment]
Thanks. Looks good on my end.

@andreyvelich ?

File: katib/installs/katib-external-db/secrets.env [hunk 3]
```diff
@@ -0,0 +1,4 @@
+KATIB_MYSQL_DB_DATABASE=
+KATIB_MYSQL_DB_HOST=
```
Comment (andreyvelich) [inline katib/installs/katib-external-db/secrets.env:2]: Should you also add `KATB_MYSQL_DB_PORT` here?

**Comment by andreyvelich** [comment]
@jlewi What do you think about Katib crds? Should it not be part of `/katib/components` also?

File: katib/installs/katib-external-db/secrets.env (hunk 3 above)
Comment (discordianfish) [reply katib/installs/katib-external-db/secrets.env:2]: Added and fixed the typo

**Comment by discordianfish** [comment]
> @jlewi What do you think about Katib crds? Should it not be part of /katib/components also?

As requested, this PR doesn't move any existing manifests, so I don't think there is any TODO for the crds. Beside that @jlewi already said it looks good from his POV. It would be great if we could minimalize the back and forth where possible, given we're all in different timezones.

**Comment by andreyvelich** [comment]
@discordianfish Alright, overall /lgtm.
Thank you for doing this!

/cc @johnugeorge @gaocegege

**Comment by jlewi** [comment]
/approved

Thanks @discordianfish!

@andreyvelich could you please create an OWNERs file under /katib so you can approve changes?

**Comment by discordianfish** [comment]
@jlewi I've just tried to test this on our cluster (by using this branch and katib-controller's repoRef/path set to `katib/installs/katib-external-db` in kfctl.yaml) but it fails with:
```
INFO[0000] Processing application: tensorboard           filename="kustomize/kustomize.go:495"
INFO[0000] Processing application: katib-crds            filename="kustomize/kustomize.go:495"
INFO[0000] Processing application: katib-controller      filename="kustomize/kustomize.go:495"
WARN[0000] Cannot get kustomization from /etc/kubeflow/kustomize/katib-controller/base: open /etc/kubeflow/kustomize/katib-controller/base/kustomization.yaml: no such file or directory  filename="kustomize/kustomize.go:714"
Error: failed to build kfApp from URI /etc/kubeflow/kfctl.yaml: couldn't generate KfApp:  (kubeflow.error): Code 500 with message: kfApp Generate failed for kustomize: Kustomize generate failed:  (kubeflow.error): Code 500 with message: couldn't generate kustomization file for component katib-controller: open /etc/kubeflow/kustomize/katib-controller/base/katib-db-manager-deployment.yaml: no such file or directory
```

I believe that's because kfctl's "unrolling" of the manifests. Not sure about the specifics but it copies the manifests from the source to /etc/kubeflow/manifests and flattens the structure, so if I specify `katib/installs/katib-external-db`, it ends up in `/etc/kubeflow/manifests/katib-controller` and the resources references end up going nowhere. So I think to make the changes here work, we need to change kfctl as well - or am I missing something and this is suppose to work already?

**Comment by andreyvelich** [comment]
> @jlewi I've just tried to test this on our cluster (by using this branch and katib-controller's repoRef/path set to `katib/installs/katib-external-db` in kfctl.yaml) but it fails with:
> 
> ```
> INFO[0000] Processing application: tensorboard           filename="kustomize/kustomize.go:495"
> INFO[0000] Processing application: katib-crds            filename="kustomize/kustomize.go:495"
> INFO[0000] Processing application: katib-controller      filename="kustomize/kustomize.go:495"
> WARN[0000] Cannot get kustomization from /etc/kubeflow/kustomize/katib-controller/base: open /etc/kubeflow/kustomize/katib-controller/base/kustomization.yaml: no such file or directory  filename="kustomize/kustomize.go:714"
> Error: failed to build kfApp from URI /etc/kubeflow/kfctl.yaml: couldn't generate KfApp:  (kubeflow.error): Code 500 with message: kfApp Generate failed for kustomize: Kustomize generate failed:  (kubeflow.error): Code 500 with message: couldn't generate kustomization file for component katib-controller: open /etc/kubeflow/kustomize/katib-controller/base/katib-db-manager-deployment.yaml: no such file or directory
> ```
> 
> I believe that's because kfctl's "unrolling" of the manifests. Not sure about the specifics but it copies the manifests from the source to /etc/kubeflow/manifests and flattens the structure, so if I specify `katib/installs/katib-external-db`, it ends up in `/etc/kubeflow/manifests/katib-controller` and the resources references end up going nowhere. So I think to make the changes here work, we need to change kfctl as well - or am I missing something and this is suppose to work already?

I think that can be because of kfctl design. @jlewi Will we use this structure for kfctl in current version or we will use it only for v3 version?

**Comment by jlewi** [comment]
/approve 

@discordianfish and @andreyvelich  I added logic to short circuit the logic in kfctl to create kustomize files.
https://github.com/kubeflow/kfctl/blob/c9afc939f04420ea822e0aa09bd9bd28c8a71b73/pkg/kfapp/kustomize/kustomize.go#L519

I think there are two cases we want to support

1. Legacy - Where kustomization.yaml is generated by kfctl
1. No magic - kustomization.yaml isn't generated by kfctl; the application in KFDef points to a valid kustomize package.

In particular, I don't think we want to mix the two; i.e. use a mix of kustomize packages.

The logic to decide which mode to use is defined here
https://github.com/kubeflow/kfctl/blob/c9afc939f04420ea822e0aa09bd9bd28c8a71b73/pkg/kfconfig/types.go#L849

Its a bit of a hack. We look for an application with a specific name. This is intended to be a "stack" which is a kustomize package which combines all the kustomize applications you want to install.
Here's the stack for GCP.
https://github.com/kubeflow/manifests/tree/master/stacks/gcp

And associated KFDef.
https://github.com/kubeflow/manifests/blob/master/stacks/examples/kfctl_gcp_stacks.experimental.yaml

So we would add the katib package to the stack kustomization.yaml and not to the KFDef.

**Comment by andreyvelich** [comment]
Thank you for your clarification @jlewi, I agree with this approach.

**Comment by discordianfish** [comment]
@jlewi So to use this, I need to migrate my kfdef to a stacks manifest? In general, I agree that this is better than having this custom kfdef but it's not trivial to migrate this.

To migrate to this, I'd like to generate all manifests both from the existing kfdef and from a new stack manifests.

I assume for the new stack based approach I can just run `kustomize build`, but for the old kfdef based approach I can't find a way to generate the manifests without applying them. `kfctl build` just generates the kustomize manifests, not the final rendered manifests. Is there a way to generate the manifests without applying them?

**Comment by jlewi** [comment]
@discordianfish looks like we are going to have to roll this back because of test breakages. #1159 is tracking rolling it forward.

>  I need to migrate my kfdef to a stacks manifest? In general, I agree that this is better than having this custom kfdef but it's not trivial to migrate this.

I guess that depends on your KFDef. There's an example KFDef for stacks here
https://github.com/kubeflow/manifests/blob/master/stacks/examples/kfctl_gcp_stacks.experimental.yaml

I wouldn't expect that to be too different from other KFDefs (E.g. just remove the GCP plugin).

> I assume for the new stack based approach I can just run kustomize build, but for the old kfdef based approach I can't find a way to generate the manifests without applying them. kfctl build 

Not quite. kfctl build is still creating the kustomization packages and not actually hydrating the manifests.

We don't currently have a command to recourse of all the directories in "${KFAPP}/kustomize".
You could write your favorite shell script to do it. That's the approach we are taking on GCP.
https://github.com/kubeflow/gcp-blueprints/blob/6f16dce8033c71dfe92d26075c62529a74d3097e/kubeflow/Makefile#L69

Or you could try running 

```
cd ${KFAPP}/kustomize
kustomize create --autodetect
```

I think that will create a kustomization.yaml file that combines all the subdirectories but I haven't tried it.

I didn't go that route because 

1. To me an uber kustomization.yaml file implies that all the packages could potentially be in the same namespace but that would actually break things 
1. Order matters in how the packages are applied and I'm not sure if an uber kustomize file preserves order.

**Pull request merged**